_posts/2021-03-10-Bootable-jar-jkube-clustering-dns-ping-openshift.adoc

```diff
@@ -0,0 +1,162 @@
+
+== Getting started
+
+The following are the key points we need to accomplish to configure our Bootable JAR cluster application to use DNS_PING:
+
+* *Define a ping service.* We need to create a https://kubernetes.io/docs/concepts/services-networking/service/#headless-services[headless service] resource, which is used by the protocol to get the IP addresses of the pods that would be part of the cluster. When we are defining this resource, we need to pay attention to the service selector and service name. The selector defines the pods that will be backed by this service. The JGroups subsystem needs to know the service name to configure the DNS_PING protocol.
+
+* *Configure the JGroups subsystem to use DNS_PING.* We need to change the default JGroups subsystem configuration of our Bootable JAR application. We will execute a CLI script at build time to configure the JGroups protocol stack, removing the default KUBE_PING protocol and adding the DNS_PING. One important detail about the DNS_PING configuration is the value we will use for the *dns_query* property. This value specifies the DNS name of the Kubernetes ping service the protocol is going to use to discover other cluster members.
+
+To sum up, we will need a Kubernetes service with a matching label selector to define which pods will be part of our cluster, and we will tell to the JGroups subsystem that we want to use such a service to locate other possible cluster members.
+
+== Define a ping service
+
+The ping service is just an additional Kubernetes service resource. We are using https://www.eclipse.org/jkube/docs/openshift-maven-plugin[JKube openshift-maven-plugin] to deploy our application, this maven plugin allows us to specify any additional resource we want to deploy. What we need to do is to create our additional resources under *src/main/jkube/* and JKube will create the resource for us when deploying the application.
+
+This is our ping service definition at https://github.com/yersan/wildfly-clustering-demo/blob/dns-ping/src/main/jkube/ping_service.yaml[src/main/jkube/ping_service.yaml]:
+----
+apiVersion: v1
+kind: Service
+metadata:
+  name: wildfly-demo-ping-service
+  annotations:
+    description: The JGroups ping service for clustering discovery.
+spec:
+  clusterIP: None
+  publishNotReadyAddresses: true
+  selector:
+    app: wildfly-clustering-demo
+----
+
+The *selector* specifies the name of our application. Without any additional modification, the name of our application is the maven group id of our project.
+We have also configured the service with *publishNotReadyAddresses: true*. This configuration specifies that when there is a DNS lookup by using this service, the result will contain all pods matching the service’s label selector, even those which are not yet in a ready state. We need it to properly discover the pods IP when they are not yet in the ready state.
```

Comment:
s/need it to/need this in order to/g

Perhaps add "We want the JGroups cluster to form before the pods are receiving end user requests."